sdk/test/opentelemetry/sdk/correlation_context/manager_test.rb

```diff
@@ -0,0 +1,190 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'test_helper'
+
+describe OpenTelemetry::SDK::CorrelationContext::Manager do
+  Context = OpenTelemetry::Context
+  let(:manager) { OpenTelemetry::SDK::CorrelationContext::Manager.new }
+
+  after do
+    Context.clear
+  end
+
+  describe '.set_value' do
+    describe 'explicit context' do
+      it 'sets key/value in context' do
+        ctx = Context.empty
+        _(manager.value('foo', context: ctx)).must_be_nil
+
+        ctx2 = manager.set_value('foo', 'bar', context: ctx)
+        _(manager.value('foo', context: ctx2)).must_equal('bar')
+
+        _(manager.value('foo', context: ctx)).must_be_nil
+      end
+    end
+
+    describe 'implicit context' do
+      it 'sets key/value in implicit context' do
+        _(manager.value('foo')).must_be_nil
+
+        Context.with_current(manager.set_value('foo', 'bar')) do
+          _(manager.value('foo')).must_equal('bar')
+        end
+
+        _(manager.value('foo')).must_be_nil
+      end
+    end
+  end
+
+  describe '.clear' do
+    describe 'explicit context' do
+      it 'returns context with empty correlation context' do
+        ctx = manager.set_value('foo', 'bar', context: Context.empty)
+        _(manager.value('foo', context: ctx)).must_equal('bar')
+
+        ctx2 = manager.clear(context: ctx)
+        _(manager.value('foo', context: ctx2)).must_be_nil
+      end
+    end
+
+    describe 'implicit context' do
+      it 'returns context with empty correlation context' do
+        ctx = manager.set_value('foo', 'bar')
+        _(manager.value('foo', context: ctx)).must_equal('bar')
+
+        ctx2 = manager.clear
+        _(manager.value('foo', context: ctx2)).must_be_nil
+      end
+    end
+  end
+
+  describe '.remove_value' do
+    describe 'explicit context' do
+      it 'returns context with key removed from correlation context' do
+        ctx = manager.set_value('foo', 'bar', context: Context.empty)
+        _(manager.value('foo', context: ctx)).must_equal('bar')
+
+        ctx2 = manager.remove_value('foo', context: ctx)
+        _(manager.value('foo', context: ctx2)).must_be_nil
+      end
+
+      it 'returns same context if key does not exist' do
```

Comment:
Is this required behaviour or an implementation detail? Should callers care?

Reply:
It's an implementation detail (optimization).